### ISSUE TITLE
Rationalize the canonical attributes

### ISSUE THREAD

**Issue by ZackButcher** [istio/istio#97]
PR https://github.com/istio/istio/pull/92 makes public our canonical attributes but during review a few issues were pointed out for specific attributes (that were out of scope of that PR):

- {source, target, origin}.namespace may be too k8s specific.
- source.port is ephemeral and we may not be able to infer the port of the container making the request.
- source.name may not be derivable, e.g. a k8s batch job. K8s does not natively have a concept of job or application either.

**Comment by ZackButcher** [comment]
@smawson if this shouldn't be in the alpha milestone push it down to wherever is appropriate.

**Comment by kyessenov** [comment]
Namespace is a general concept, but the way k8s uses it for isolation and without nesting, is fairly specific to k8s. I think we should just use FQ service names and not have namespaces. This should generalize to federated services, services with custom domain names, and perhaps, raw VMs.

**Comment by kyessenov** [comment]
Regarding `source.name`: pods can be late bound to services/batch jobs/etc. That means proxy may not know ahead of time what services it belongs to or what batch job it is executing.

**Comment by mandarjog** [comment]
We are not mandating that `source.name` be set by the proxy.
As long as proxy sets `source.uid`   and mixer knows how to extract other metadata from it, we are ok.

We should add a "source" column to attributes.md. It should indicate which component produces this attribute. `source.uid` is proxy and `source.name` should be mixer. 

@kyessenov does proxy need to know all pod attributes (like service membership, labels) ahead of time?  Proxy manager is watching the cluster for changes, are these changes not being watched for?

**Comment by kyessenov** [comment]
Proxy manager is only watching for services (kubernetes kind `Service). Hence, the confusion why we dropped `source.service`. I do not think we should be watching for all k8s resources, more so, not in every pod.

**Comment by kyessenov** [comment]
@mandarjog this is conflicting information. All these attributes are supposed to be produced by the proxy, not derived by mixer.

**Comment by ZackButcher** [comment]
Not quite: these are the canonical attributes in an Istio deployment, not the set of attributes produced by the proxy. The difference is because the mixer is also able to synthesize facts. The list of attributes in this repo is a superset of the manifests of any individual component of an istio deployment.

**Comment by kyessenov** [comment]
It would be useful to have a separate list of attributes that need to be produced by the proxy.

**Comment by mandarjog** [comment]
Yes, which is why we need "source" column as in my previous comment. [https://github.com/istio/istio/issues/97#issuecomment-290551567](https://github.com/istio/istio/issues/97#issuecomment-290551567)

A policy author needs to know the union of attributes which is this document.
We really don't __need__ a manifest for alpha release. This document is the authority, proxy should produce the attributes it is asked to produce and mixer will do the same.

manifest comes into play post alpha or when we are adding a new proxy in the istio deployment.

**Comment by smawson** [comment]
Ideally we would use the manifest to validate configuration in Alpha, is that something we can make happen? I.e. if I type source.nmae I'll get an error during configuration instead of just having my policy be broken at runtime.

**Comment by mandarjog** [comment]
That will happen for alpha. The part that may not happen is proxy declaring it's own manifest and the ability to collate all manifests in a `vocabulary`.

We already have a `vocabulary` doc that the proxy we are writing will support. And we can validate against that.

**Comment by smawson** [comment]
Copying comment from @frankbu over:

It would seem totally reasonable for source.service to have the value nil or "none" when the caller is not a service. But it still lets one identify calls that are from a particular service. Maybe source.name is intended for this purpose? I'm not sure I'm understanding what value source.name is intended to have when the call is from a service.

Say I want to rate-limit calls to the ratings service from the reviews service test version. I would have imagined something like selector: target.service == "ratings.*" && source.service == "reviews.*" && "test" in source.labels. Is something like this not what you guys are imagining?

**Comment by smawson** [comment]
Basically the problem is with the term "service", which we need to be careful with. In the current attribute vocab, "service" means "the name that was used to resolve the destination of the call".  I didn't think it made sense to include a source service, since in many cases a source has no service, or could have multiple. But in maybe the majority of cases a source would actually be part of one and only one service, and maybe in those cases source.service is actually meaningful and useful?

I think the usage of service in frank's example is more like what was supposed to be meant by source.name and target.name, something like application name or deployment name. I'd expect rules to be written like target.service == "ratings" && source.name == "reviews".

Basically we don't have service->service, we have application->service.  We need to figure out a clean way to describe this and make it understandable, and especially explain exactly why the distinction is needed.

**Comment by geeknoid** [comment]
Note that the canonical list of attributes should now be considered to be on http://istio.io/docs/reference/attribute-vocabulary.html which can be edited from http://github.com/istio/istio.github.io

**Comment by linsun** [comment]
@ZackButcher what is remaining for this?  i saw no activity for 4+ months.

**Comment by ldemailly** [comment]
I think this is done as far as 0.2 is concerned. closing. reopen if I missed something.
